### PR TITLE
New alignment option: `join='strict'`

### LINKED ISSUE #6806: New alignment option: "exact" without broadcasting OR Turn off automatic broadcasting
### Is your feature request related to a problem?

If we have two objects with dims `x` and `x1`, then `xr.align(..., join="exact")` will pass because these dimensions are broadcastable.

I'd like a stricter option (`join="strict"`?) that disallows broadcasting.

### Describe the solution you'd like

```python
xr.align(
    xr.DataArray([1], dims="x"),
    xr.DataArray([1], dims="x1"),
    join="strict",
)
```
would raise an error.

It'd be nice to have this as a built-in option so we can use
``` python
with xr.set_options(arithmetic_join="strict"):
    ...
```

### Describe alternatives you've considered

An alternative would be to allow control over automatic broadcasting through the `set_options` context manager., but that seems like it would be more complicated to implement.

### Additional context

This turns up in staggered grid calculations with xgcm where it is easy to mistakenly construct very high-dimensional arrays because of automatic broadcasting.

### LINKED ISSUE #8231: xr.concat concatenates along dimensions that it wasn't asked to
### What happened?

Here are two toy datasets designed to represent sections of a dataset that has variables living on a staggered grid. This type of dataset is common in fluid modelling (it's why xGCM exists).

```python
import xarray as xr

ds1 = xr.Dataset(
    coords={
        'x_center': ('x_center', [1, 2, 3]),
        'x_outer':  ('x_outer',  [0.5, 1.5, 2.5, 3.5]),  
    },
)

ds2 = xr.Dataset(
    coords={
        'x_center': ('x_center', [4, 5, 6]),
        'x_outer':  ('x_outer',  [4.5, 5.5, 6.5]),  
    },
)
```

Calling `xr.concat` on these with `dim='x_center'` happily concatenates them
```python
xr.concat([ds1, ds2], dim='x_center')
```
```
<xarray.Dataset>
Dimensions:   (x_outer: 7, x_center: 6)
Coordinates:
  * x_outer   (x_outer) float64 0.5 1.5 2.5 3.5 4.5 5.5 6.5
  * x_center  (x_center) int64 1 2 3 4 5 6
Data variables:
    *empty*
```
but notice that the returned result has been concatenated along *both* `x_center` and `x_outer`.

### What did you expect to happen?

I did not expect this to work. I definitely didn't expect the datasets to be concatenated along a dimension I didn't ask them to be concatenated along (i.e. `x_outer`).

What I expected to happen was that (as by default `coords='different'`) both variables would be attempted to be concatenated along the `x_center` dimension, which would have succeeded for the `x_center` variable but failed for the `x_outer` variable. Indeed, if I name the variables differently so that they are no longer coordinate variables then that is what happens:

```python
import xarray as xr

ds1 = xr.Dataset(
    data_vars={
        'a': ('x_center', [1, 2, 3]),
        'b':  ('x_outer',  [0.5, 1.5, 2.5, 3.5]),  
    },
)

ds2 = xr.Dataset(
    data_vars={
        'a': ('x_center', [4, 5, 6]),
        'b':  ('x_outer',  [4.5, 5.5, 6.5]),  
    },
)
```
```python
xr.concat([ds1, ds2], dim='x_center', data_vars='different') 
```
```
ValueError: cannot reindex or align along dimension 'x_outer' because of conflicting dimension sizes: {3, 4}
```

### Minimal Complete Verifiable Example

_No response_

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

_No response_

### Anything else we need to know?

I was trying to create an example for which you would need the automatic combined concat/merge that happens within `xr.combine_by_coords`.

### Environment

xarray `2023.8.0`

### PR DESCRIPTION
**Pull request by etienneschalk**
Title: New alignment option: `join='strict'`

- [ ] Closes #8231
- [x] Closes #6806
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
    - [x] What's new entry   
    - [x] Refer to PR ID (cannot be done before the PR has been created) 
- [x] New functions/methods are listed in `api.rst`
    - No new functions/methods.

## Motive

This PR is motivated by solving of the following issues:

- xr.concat concatenates along dimensions that it wasn't asked to #8231
	- New alignment option: "exact" without broadcasting OR Turn off automatic broadcasting #6806

**The current PR does not solve the unexpected issue described in #8231 without a change in user-code**. Indeed, in the tests written, it is shown that to get the said expected behavior, the user would have to use the new `join='strict'` mode suggested in #6806 for the concatenation operation. Only in that case, the uniqueness of the indexed dimensions' names will be checked, re-using the same logic that was already applied for `join='override'` in `Aligner.find_matching_indexes` 

This may not be enough to fix #8231. If that isn't, I can split the PR into two, first one for adding the `join='strict'` for #6806 and later on one for #8321. 

## Technical Details 

I try to detail here my thought process. Please correct me if there is anything wrong. This is my first time digging into this core logic! 

Here is my understanding of the terms:

- An **indexed dimension** is attached to a coordinate variable
- An **unindexed dimension** is not attached to a coordinate variable (_"Dimensions without coordinates"_)

Input data for Scenario 1, tested in `test_concat_join_coordinate_variables_non_asked_dims`

```python
    ds1 = Dataset(
        coords={
            "x_center": ("x_center", [1, 2, 3]),
            "x_outer": ("x_outer", [0.5, 1.5, 2.5, 3.5]),
        },
    )

    ds2 = Dataset(
        coords={
            "x_center": ("x_center", [4, 5, 6]),
            "x_outer": ("x_outer", [4.5, 5.5, 6.5]),
        },
    )
```

Input data for Scenario 2, tested in `test_concat_join_non_coordinate_variables`

```python
    ds1 = Dataset(
        data_vars={
            "a": ("x_center", [1, 2, 3]),
            "b": ("x_outer", [0.5, 1.5, 2.5, 3.5]),
        },
    )

    ds2 = Dataset(
        data_vars={
            "a": ("x_center", [4, 5, 6]),
            "b": ("x_outer", [4.5, 5.5, 6.5]),
        },
    )
```

Logic for non-indexed dimensions logic was working "as expected", as it relies on `Aligner.assert_unindexed_dim_sizes_equal`, checking that unindexed dimension sizes are equal as its name suggests. (Scenario 1)

However, the logic for indexed dimensions was surprising as such an expected check on dimensions' sizes was not performed. A check exists in `Aligner.find_matching_indexes` but was only applied to `join='override'`. Applying it for `join='strict'` too is suggested in this Pull Request.

### CODE REVIEW

**Comment by etienneschalk** [comment]
CI Failed because of multiple timeouts being > 180s:

https://github.com/pydata/xarray/actions/runs/7768517575/job/21186540956?pr=8698

Does it happen sometimes? If so, is it possible to re-trigger the failed pipeline?

Thanks!

File: xarray/tests/test_concat.py [hunk 1]
```diff
@@ -1261,3 +1262,103 @@ def test_concat_index_not_same_dim() -> None:
...
+        [ds1, ds2],
+        dim="x_center",
+        data_vars="different",
+        coords="different",
+        join="outer",
+    )
+    assert all(actual_xds == expected_wrongly_concatenated_xds)
+
+    # Using join='strict'
+    # A check similar to the one made on non-indexed dimensions regarding their sizes.
```
Comment (max-sixty) [inline xarray/tests/test_concat.py:1301]: Great explanation, thanks.

What's the difference between `strict` and `exact`? Could we show the difference?

---

Assuming we do merge this, we'll want this sort of explanation in the docs...

**Review by max-sixty — COMMENTED** [review]
Sorry this didn't get an earlier review...

I left one question. 

Others know more about this area so they should feel free to comment; and it's a big enough change that I'll ask others to approve before merging

File: xarray/tests/test_concat.py (hunk 1 above)
Comment (etienneschalk) [reply xarray/tests/test_concat.py:1301]: > What's the difference between strict and exact? Could we show the difference?

**TLDR: join='exact` does not consider dimension names when aligning. join='strict' expects matching dimension names when aligning.**

The best difference is exposed in [`test_align_exact_vs_strict`](https://github.com/pydata/xarray/pull/8698/files#diff-ca5c2de2fe6e9e25fbf22bd53e4976c15da74900dfb14deb7e6e87f5377230e3R3197) . We can see an instance of two DataArrays  whom alignment is accepted with join='exact', but not with join='strict'. 'exacts' allows _shape-compatible_ DataArrays to be aligned with differing dimension names, whereas 'strict' forbids.

To be fair, I would have expected that this more restrictive logic would just have been added to the existing 'exact' mode. But when I tried so, many tests were failing. Also, I only tested on toy examples. I was quite unfamiliar with this alignment logic so I sticked to following the desired behaviour described in https://github.com/pydata/xarray/issues/6806 ; and trusted the existing test coverage for the rest.

I would say the biggest addition is this new check that is performed early during the aligment operation:

https://github.com/pydata/xarray/pull/8698/files#diff-c32c8ede32701deaba092f13d721242d5ef69144b2d5e76ecf1748333f1106f5R595

It is only applied for the `strict` mode, and checks that all objects to be aligned have the exact same dimensions. It is really a "symbolic" check on the dimension names, to ensure that compatible arrays with same shape but different dimension names will not _slip through the cracks_. It is pretty restrictive and add more constraints, so that's probably be why this mode would be an opt-in as it could annoy some users.

File: xarray/tests/test_dataarray.py [hunk 2]
```diff
@@ -3194,6 +3194,26 @@ def test_align_str_dtype(self) -> None:
...
         assert expected_b.x.dtype == actual_b.x.dtype
 
+    def test_align_exact_vs_strict(self) -> None:
+        xda_1 = xr.DataArray([1], dims="x1")
+        xda_2 = xr.DataArray([1], dims="x2")
+
+        # join='exact' passes
+        aligned_1, aligned_2 = xr.align(xda_1, xda_2, join="exact")
+        assert aligned_1 == xda_1
+        assert aligned_2 == xda_2
```
Comment (max-sixty) [inline xarray/tests/test_dataarray.py:3204]: The existing behavior does seem surprising.

Is it only an issue for arrays with 1 element though? Are there other examples of this behavior?

(Another option would be refining `exact` if that's not the behavior we want...)

File: xarray/tests/test_concat.py (hunk 1 above)
Comment (max-sixty) [reply xarray/tests/test_concat.py:1301]: Thanks a lot. I added another question to that example.

File: xarray/tests/test_dataarray.py [hunk 3]
```diff
@@ -3206,14 +3206,36 @@ def test_align_exact_vs_strict(self) -> None:
...
-                r"([('x1',), ('x2',)])"
+            match=re.escape(
+                "cannot align objects with join='strict' "
+                "because given objects do not share the same dimension names "
+                "([('x1',), ('x2',)])"
             ),
         ):
             xr.align(xda_1, xda_2, join="strict")
 
+    def test_align_exact_vs_strict_2d(self) -> None:
```
Comment (etienneschalk) [inline xarray/tests/test_dataarray.py:3217]: Hello @max-sixty 

I added a new test: aligning three 2D arrays with differing dims, following https://github.com/pydata/xarray/pull/8698#discussion_r1484737785

File: xarray/tests/test_dataarray.py (hunk 2 above)
Comment (etienneschalk) [reply xarray/tests/test_dataarray.py]: > The existing behavior does seem quite surprising. Is it only an issue with 1D arrays?

I added a test with 2D arrays, they are concerned too

What I can do as a simple test is to systematically transform 'exact' to 'strict' when entering the alignment logic and pushing this logic as a separate draft PR ("Reinforce alignment checks when join='strict'") to see the failing tests.

**Comment by max-sixty** [comment]
OK, I think I see what's going on now! Sorry I was slow. I appreciate you doing so much to make it easy to engage with the issue.

So when we say "'exact' allows _shape-compatible_ DataArrays to be aligned with differing dimension names, whereas 'strict' forbids.", the existing behavior doesn't align dimensions with different names — it just ignores them.

My mental model has that as correct behavior — I haven't observed much demand for "only align these arrays if _all_ the dimensions match"

Is that consistent with your view?

The change I would support is https://github.com/pydata/xarray/issues/6806, which would make `strict` raise an error if one object has a dimension with size 1 and another has a dimension with size `n`. Currently `exact` allows that.

Does that make sense? Lmk if I'm still missing something.

File: xarray/tests/test_concat.py [hunk 4]
```diff
@@ -1261,3 +1262,103 @@ def test_concat_index_not_same_dim() -> None:
...
+    # Not using strict mode will allow the concatenation to surprisingly happen
+    # even if `x_outer` sizes do not match
+    actual_xds = concat(
+        [ds1, ds2],
+        dim="x_center",
+        data_vars="different",
+        coords="different",
+        join="outer",
+    )
+    assert all(actual_xds == expected_wrongly_concatenated_xds)
```
Comment (max-sixty) [inline xarray/tests/test_concat.py:1298]: FYI for these sorts of comparisons, check out `assert_identical` (or `assert_equal`)

**Comment by etienneschalk** [comment]
Hello,

Thank your for your time giving me feedback on this PR!

⁂

> the existing behavior doesn't align dimensions with different names — it just ignores them.

Yes, they are not considered. It is like manipulating raw numpy arrays: the typing is _structural_ and not _nominal_, I would say. This situation reminds me of the structural and nominal terminology used to describe type systems. So even if the two DataArrays are _structurally_ equivalent, the 'strict' mode would fail because of the difference in _names_. Definitely the Python community is more leaning towards structural typing: operations on numpy arrays rely on the structure of arrays, [PEP 544 – Protocols: Structural subtyping (static duck typing)](https://peps.python.org/pep-0544/), etc. 

But sometimes, having the safety of nominal typing can be useful. In a way, this is what xarray seems to aim to address by naming dimensions. Indeed, it is common to get confused with `x, y, row, col, latitude, longitude` conventions when working with georeferenced rasters, for instance. Structure alone is in such cases insufficient to work confidently with data, and xarray helps with that.

I found this interesting quote on the Wikipedia page of [Structural type system](https://en.wikipedia.org/wiki/Structural_type_system):

> A pitfall of structural typing versus nominative typing is that two separately defined types intended for different purposes, but accidentally holding the same properties (e.g. both composed of a pair of integers), could be considered the same type by the type system, simply because they happen to have identical structure. One way this can be avoided is by creating one [algebraic data type](https://en.wikipedia.org/wiki/Algebraic_data_type "Algebraic data type") for each use.

This issue is also well-known in the TypeScript world. See [How do I prevent two types from being structurally compatible?](https://github.com/Microsoft/TypeScript/wiki/FAQ#how-do-i-prevent-two-types-from-being-structurally-compatible)

So to summarize, my mental model is the following:

- 'exact' is structural
- 'strict' is nominal

⁂

> The change I would support is [#6806](https://github.com/pydata/xarray/issues/6806), which would make `strict` raise an error if one object has a dimension with size 1 and another has a dimension with size `n`. Currently `exact` allows that.

I understood that the issue described in https://github.com/pydata/xarray/issues/6806#issue-1308371056 was related to arrays having the same structure but different dimension names being successfully aligned, not related to dimension sizes.

Also, I understand that implementing an option to use 'strict' by default would solve the scenario described in https://github.com/pydata/xarray/issues/6806#issuecomment-1188339194 (not implemented in this PR)

I added new test to experiment. Here is a test matrix result of the current implementation:

| `test_align_exact_vs_strict_*` | dim names | Nominally equivalent (same dim names) | dim sizes | Structurally equivalent (same dim sizes) | exact | strict | Error message |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| `*_same_dim_same_size` | x, x | Yes | 1, 1 | Yes | :ok: | :ok: | - |
| `*_same_dim_differing_sizes` | x, x | Yes | 1, 2 | No | :x: | :x: | (E1) |
| `*_differing_dims_same_sizes` | x1, x2 | No | 1, 1 | Yes | :ok: | :x: | (E2) |
| `*_differing_dims_differing_sizes` | x1, x2 | No | 1, 2 | No | :ok: | :x: | (E2) |

(E1): Structural error 
```
"cannot reindex or align along dimension 'x' because of "
"conflicting dimension sizes: {1, 2}"
```

(E2): Nominal error (check happens before the structural error has a chance to happen)
```
"cannot align objects with join='strict' "
"because given objects do not share the same dimension names "
"([('x1',), ('x2',)])"
```

Sorry, this starts to become very long!

**Comment by max-sixty** [comment]
(Great! I asked a question at #6806 to clarify the request)

File: doc/whats-new.rst [hunk 5]
```diff
@@ -23,12 +23,18 @@ v2024.02.0 (unreleased)
...
+  By `Etienne Schalk <https://github.com/etienneschalk>`_.
 - Added a simple `nbytes` representation in DataArrays and Dataset `repr`.
   (:issue:`8690`, :pull:`8702`).
   By `Etienne Schalk <https://github.com/etienneschalk>`_.
 - Allow negative frequency strings (e.g. ``"-1YE"``). These strings are for example used
   in :py:func:`date_range`,  and :py:func:`cftime_range` (:pull:`8651`).
   By `Mathias Hauser <https://github.com/mathause>`_.
+- Added a ``join='strict'`` mode for ``Aligner.align`` and related classes.
+  (:issue:`7132`, :issue:`8230`).
+  By `Etienne Schalk <https://github.com/etienneschalk>`_.
```
Comment (dcherian) [inline doc/whats-new.rst:37]:
```suggestion
```

File: xarray/core/alignment.py [hunk 6]
```diff
@@ -264,13 +276,14 @@ def find_matching_indexes(self) -> None:
...
+        if self.join == "override" or not self.broadcast:
             for dim_sizes in all_indexes_dim_sizes.values():
                 for dim, sizes in dim_sizes.items():
                     if len(sizes) > 1:
                         raise ValueError(
-                            "cannot align objects with join='override' with matching indexes "
-                            f"along dimension {dim!r} that don't have the same size"
+                            f"cannot align objects with join={self.join!r} or "
+                            f"broadcast={self.broadcast!r} with matching indexes "
+                            f"along dimension {dim!r} that don't have the same size ({sizes!r})"
```
Comment (dcherian) [inline xarray/core/alignment.py:286]: Perhaps this is clearer:
```suggestion
                            f"cannot align objects"
                            f"with indexes "
                            f"along dimension {dim!r} that don't have the same size ({sizes!r}) when {message} "
```

where

```
message = "join='override'" if self.join == "override" else "broadcast=False`
```

File: xarray/core/alignment.py [hunk 7]
```diff
@@ -478,6 +491,20 @@ def assert_unindexed_dim_sizes_equal(self) -> None:
                     f"because of conflicting dimension sizes: {sizes!r}" + add_err_msg
                 )
 
+    def assert_equal_dimension_names(self) -> None:
+        # When broadcasting is disabled, only allows objects having the exact same dimensions' names.
+        if self.broadcast:
+            return
+
+        unique_dims = set(tuple(o.sizes) for o in self.objects)
```
Comment (dcherian) [inline xarray/core/alignment.py:499]:
```suggestion
        unique_dims = set(tuple(o.dims) for o in self.objects)
```

This seems a little clearer

File: xarray/core/alignment.py [hunk 8]
```diff
@@ -576,7 +604,7 @@ def align(self) -> None:
 
         if self.join == "override":
             self.override_indexes()
-        elif self.join == "exact" and not self.copy:
+        elif not self.copy and self.join == "exact":
```
Comment (dcherian) [inline xarray/core/alignment.py:607]: Nothing has changed here? It's usually nicer to minimize the diff as much as possible, that makes reviewing easier.

File: xarray/core/alignment.py [hunk 9]
```diff
@@ -710,7 +745,8 @@ def align(
         - "override": if indexes are of same size, rewrite indexes to be
           those of the first object with that dimension. Indexes for the same
           dimension must have the same size in all objects.
-
+    broadcast : bool
+        The alignment fails if dimensions' names differ.
```
Comment (dcherian) [inline xarray/core/alignment.py:749]:
```suggestion
    broadcast : bool, optional
        Turn off automatic broadcasting of all objects along dimensions that are present in some but not all objects.
        If False, this will raise an error when all objects do *not* have the same dimensions.
```

File: xarray/core/computation.py [hunk 10]
```diff
@@ -969,6 +974,8 @@ def apply_ufunc(
         dimensions as input and vectorize it automatically with
         :py:func:`numpy.vectorize`. This option exists for convenience, but is
         almost always slower than supplying a pre-vectorized function.
+    broadcast : bool
+        The alignment fails if dimensions' names differ.
```
Comment (dcherian) [inline xarray/core/computation.py:978]:
```suggestion
    broadcast : bool, optional
        Turn off automatic broadcasting of all objects along dimensions that are present in some but not all objects.
        If False, this will raise an error when all objects do *not* have the same dimensions.
```

File: xarray/core/alignment.py [hunk 11]
```diff
@@ -886,6 +923,7 @@ def align(
 def deep_align(
     objects: Iterable[Any],
     join: JoinOptions = "inner",
+    broadcast: bool = True,
```
Comment (dcherian) [inline xarray/core/alignment.py:926]: This needs to be forwarded to `align`.

File: xarray/core/computation.py [hunk 12]
```diff
@@ -969,6 +974,8 @@ def apply_ufunc(
         dimensions as input and vectorize it automatically with
         :py:func:`numpy.vectorize`. This option exists for convenience, but is
         almost always slower than supplying a pre-vectorized function.
+    broadcast : bool
```
Comment (dcherian) [inline xarray/core/computation.py:977]: I think we should skip `apply_ufunc` for now (and just ignore `arithmetic_broadcast` in `dot`)

Seems like there are some confusing implications to implementing it. See https://github.com/pydata/xarray/issues/1618 for an alternative approach.

File: xarray/core/options.py [hunk 13]
```diff
@@ -91,26 +94,35 @@ def _positive_integer(value: int) -> bool:
     return isinstance(value, int) and value > 0
 
 
+def _is_boolean(value: Any) -> bool:
+    return isinstance(value, bool)
```
Comment (dcherian) [inline xarray/core/options.py:98]: minor comment: This is a matter of taste but this feels like a fairly tiny benefit for the added indirection. The diff would be a lot smaller too, if you just followed the existing copy-paste approach.

File: xarray/tests/test_dataarray.py [hunk 14]
```diff
@@ -3203,6 +3204,118 @@ def test_align_str_dtype(self) -> None:
...
+        xda_1 = xr.DataArray([1], dims="x")
+        xda_2 = xr.DataArray([1], dims="x")
+
+        aligned_1, aligned_2 = xr.align(xda_1, xda_2, join="exact", broadcast=True)
+        assert_identical(aligned_1, xda_1)
+        assert_identical(aligned_2, xda_2)
+
+        aligned_1, aligned_2 = xr.align(xda_1, xda_2, join="exact", broadcast=False)
+        assert_identical(aligned_1, xda_1)
+        assert_identical(aligned_2, xda_2)
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:3217]:
```suggestion
    @pytest.mark.parametrize("broadcast", [True, False])
    def test_broadcast_on_vs_off_same_dim_same_size(self, broadcast) -> None:
        xda = xr.DataArray([1], dims="x")

        aligned_1, aligned_2 = xr.align(xda, xda, join="exact", broadcast=broadcast)
        assert_identical(aligned_1, xda)
        assert_identical(aligned_2, xda)
```

File: xarray/tests/test_concat.py [hunk 15]
```diff
@@ -1261,3 +1262,90 @@ def test_concat_index_not_same_dim() -> None:
...
+        concat(
+            [ds1, ds2],
+            join="exact",
+            dim="x_center",
+            data_vars="different",
+            coords="different",
+        )
+
+
+@pytest.mark.parametrize("join", ("outer", "exact"))
```
Comment (dcherian) [inline xarray/tests/test_concat.py:1319]: lets merge this with the previous test.

Comment (etienneschalk) [reply xarray/tests/test_concat.py:1319]: They are not exactly the same, the previous test `test_concat_join_coordinate_variables_non_asked_dims` tests coordinate variables while this one `test_concat_join_non_coordinate_variables` tests non-coordinate variables, also the parametrization is not possible with the previous test as the behaviour differs when using join=outer vs join=exact for coordinate variables

File: xarray/core/alignment.py (hunk 8 above)
Comment (etienneschalk) [reply xarray/core/alignment.py]: This is definitely a leftover for when I was tinkering with the logic!

File: xarray/core/options.py (hunk 13 above)
Comment (etienneschalk) [reply xarray/core/options.py:98]: I tried to apply "opportunistic refactoring", since I was updating this zone of the code. Currently, there is a mixed approach: lambdas are used, but a function `_positive_integer` already exists. So I tried to align all the existing checks to `_positive_integer`. It looks at least to me a little bit cleaner to refer multiple times one function rather than define multiple duplicated lambdas.
I understand however the main downside of it adding unrelated noise to the diff, sorry for that. But now that the change is made, should I revert it or keep it?

File: xarray/core/computation.py [hunk 16]
```diff
@@ -282,6 +282,7 @@ def apply_dataarray_vfunc(
     *args,
     signature: _UFuncSignature,
     join: JoinOptions = "inner",
+    broadcast: bool = True,
```
Comment (dcherian) [inline xarray/core/computation.py:285]: Sorry for being unclear (again). I don't think we need this in `align` at all.

We should simply be checking `OPTIONS["arithmetic_broadcast"]` in `Variable._binary_op` . The whole business with `align` was a misunderstanding of mine.

Comment (etienneschalk) [reply xarray/core/computation.py:285]: I will close this PR as it diverged too much from the original wanted behavior
